Actor/Movement controller refactor

dmg9626 opened the pull request:
## Changes:

### Movement:
- refactored movement for each perspective to be handled in a non-monobehavior Movement class
    - Follow movement is also done this way (see FollowMovement.cs)
- moved animation code to its own AnimationController.cs class
- implemented sprinting (hold SHIFT to sprint)
    - scales base movement speed with sprint multiplier

### Actors:
- Both Oliver and Cat have been reworked and turned into Actors!
- Each now use the Actor component for movement/animation/interaction stuff, while PlayerController just selects one Actor to control.
    - When player is controlling Oliver, Cat gets its movement input from FollowMovement
- I did this so the PlayerController can easily switch between controlling Oliver and Cat - just point to the other actor and assign the other one to use FollowMovement to follow you 
    - We can also just create a NoMovement class that always returns Vector3.zero so the actor stays in place

### PlayerController class:
- **BIG BOY** refactor time here 
- Movement, animation, and control in general has been mostly moved into different classes (Actor, AnimationController, Movement, etc)
- Basically just drives an Actor instance (whichever character player is controlling)

Switching between Oliver and Cat isn't implemented yet, I just wanted to get these big structural changes in now before I make further changes. Also now Victor can work on the animation stuff without bumping into me too much.


# TLDR:
- Big changes to movement
    - Different Movement classes for each camera perspective
    - Actor receives movement from a Movement class
- Every character (player-controlled or not) that can move around/needs animation will use an Actor component

Test changes in the ukiyoe scene - should work about the same as it did before, but without animation for oliver/cat

*Update: I also organized the FUCK out of the main scene yo CHECK that shit OUT*

---

## What to Test:
- Everything!
    - I changed a whole bunch of code that touched lots of different features, so there's a good chance I broke a handful of things
    - I tried to test as much as I could, but I'm sure I missed something
    - Take a good thorough pass of every puzzle/feature in our game!
- Note: Check the **Known Bugs** section below before testing


#### Known Bugs:

- ~~Animation for Oliver/Cat is no longer working, needs to be fixed to conform to new movement code~~ (FIXED)
- ~~Sprite doesn't always *quite* face the camera in orthographic perspectives, due to camera rotation in 3D perspective~~ (FIXED)
- Proper animations not working yet for Top-Down perspective, or an Actor following the player
- ~~Player movement no longer frozen when interacting with puzzles/dialogue/etc~~ (FIXED)
- ~~Pressing E when not near an interactable puzzle broke player controls ("No cameras rendering")~~ (FIXED)

Commit:
moved interaction code to separate method; added comments

Commit:
applied ukiyo-e scene changes to oliver prefab - adjusted collider scale, movement speed, camera settings, etc.

Commit:
refactored interaction code - delegated to a PlayerInteractionController component for now, to keep it out of the way during my refactor

Commit:
added comments to interaction controller

Commit:
Merge branch 'develop' into player-controller-redesign

Commit:
created abstract Movement class with concrete implementation of follow behavior (FollowMovement)

Commit:
refactored player movement to implementations of Movement class (one for each camera perspective)

Commit:
deleted old code, added comments

Commit:
updated Oliver prefab - applied scene changes to main prefab

Commit:
increased oliver/cat movement speed

Commit:
moved animation code for Oliver/Cat to an AnimationController component - victor can pick up the rest from here i think

Commit:
cleaned up GameManager a bit

Commit:
added TODOs to code; forgot to include prefab overrides in previous commit

Commit:
(hopefully) resolved merge conflicts with develop; Orb seemed to disappear from oliver, so I added one back to his prefab

Commit:
hid orb on Oliver, added missing PlayerController (idk why this got lost???)

Commit:
changed Movement to be a ScriptableObject, so we can manage separate instances of them (could be useful for making movement a little different between Oliver/Cat)

Commit:
- created Actor.cs class to handle generalized actor behavior (interaction, movement, animation, etc)
- PlayerController now controls an actor, which can (in theory) be changed at runtime between Oliver and Cat
- PlayerController is now a Singleton
- changed other code referencing PlayerController to point to its currently-controlled actor when necessary

Commit:
FollowMovement now has followPlayer setting - sets target to Player if true (useful when not calling constructor)

Commit:
defensive movement checking on Actor.Start()

Commit:
removed spaces from movement scriptobj filenames

Commit:
fixed movement speed bug: PlayerController was calling Actor.Move(), doubling its movement speed

Commit:
recreated Oliver and Cat as Actors - after adding cameras to Cat we should be able to swap control betwen the two

Commit:
added/updated comments

Commit:
fixed movement bug - [SerializeField] was commented out in Actor.cs, breaking serialized connection to Actor.movemen

Commit:
merged changes from develop:

- took scene changes from develop
- commented out code in PerspectiveManager - seems to be repeating functionality from old GameManager.cs script
- changed some references to PlayerController to use Actor component instead (just enough to fix compile errors)

Commit:
updated scene (added PlayerController component to GameManager obj, changed components on/references to Oliver's Actor component, enough to fix runtime errors when moving around)

Commit:
organized koi fish pond objects into a proper hierarchy

Commit:
organized object hierarchy for zodiac lanterns

Commit:
organized tree object hierarchy

Commit:
- organized zodiac/liars/teledoor puzzle hierarchy (grouped/disabled obsolete teledoor stuff, should delete later when confident)
- deleted duplicate instance of red lantern

Commit:
organized object hierarchy in Ukiyoe House

Commit:
organized all environmental/decorative objects under single parent object Environment

Commit:
got animations working kinda, but they're not facing in the direction of (relative) player movement

Commit:
added comments to previous code changes

Commit:
changed Actor.ghostCamera reference to be of type PerspectiveCameraControl, rather than just a GameObject (cuts down on GetComponent<> calls)

Commit:
- got animations mostly working: TopDown doesn't work yet, and Perspective doesn't work well with diagonal movement

Commit:
removed old movement code (FollowMoveController.cs, MoveController.cs)

Commit:
actor sprites now always face towards camera, regardless of perspective

Commit:
updated comments. cleaned up code a bit

Commit:
changed Movement.GetMovement() to take an Actor instead of Transform as argument

Commit:
changed idle/walk/run value thresholds in Oliver and Cat animation controllers

Commit:
Merge branch 'develop' into actor-controller-refactor

cptvictor commented on 2D3D_UnityProject/Assets/Scripts/GameManager.cs:
```diff
@@ -24,36 +31,45 @@ public class GameManager : MonoBehaviour {
     void Update () {
-        if (Input.GetKeyDown ("up")) {
+        Actor playerActor = playerControl.GetActor();
+        if (Input.GetKeyDown (KeyCode.UpArrow)) {
```
```suggestion
        if (Input.GetKeyDown (KeyCode.UpArrow)) 
        {
```

cptvictor commented on 2D3D_UnityProject/Assets/Scripts/GameManager.cs:
```diff
@@ -24,36 +31,45 @@ public class GameManager : MonoBehaviour {
             perspControl.enabled = false;
-        } else if (Input.GetKeyDown ("right")) {
+        } 
+        else if (Input.GetKeyDown (KeyCode.RightArrow)) {
```
```suggestion
        else if (Input.GetKeyDown (KeyCode.RightArrow)) 
        {
```

cptvictor commented on 2D3D_UnityProject/Assets/Scripts/GameManager.cs:
```diff
@@ -24,36 +31,45 @@ public class GameManager : MonoBehaviour {
             perspControl.enabled = false;
-        } else if (Input.GetKeyDown ("left")) {
+        } 
+        else if (Input.GetKeyDown (KeyCode.LeftArrow)) {
```
```suggestion
        else if (Input.GetKeyDown (KeyCode.LeftArrow)) 
        {
```

cptvictor commented on 2D3D_UnityProject/Assets/Scripts/GameManager.cs:
```diff
@@ -24,36 +31,45 @@ public class GameManager : MonoBehaviour {
-            playerControl.constrainY (false);
-        } else if (Input.GetKeyDown ("down")) {
+        } 
+        else if (Input.GetKeyDown (KeyCode.DownArrow)) {
```
```suggestion
        else if (Input.GetKeyDown (KeyCode.DownArrow)) 
        {
```

cptvictor commented on 2D3D_UnityProject/Assets/Scripts/Player/PlayerController.cs:
```diff
@@ -2,143 +2,31 @@
-                y = jumpSpeed;
-            }
+        // Make sure we have an actor
+        if(actor == null) {
```
```suggestion
        if(actor == null) 
        {
```

cptvictor commented on 2D3D_UnityProject/Assets/Scripts/Player/Actor.cs:
```diff
@@ -0,0 +1,117 @@
+
+        // TODO: set default movement scheme and camera perspective in the same place
+        // Make sure there's a movement type specified
+        if(movement == null) {
```
```suggestion
        if(movement == null) 
        {
```

cptvictor commented on 2D3D_UnityProject/Assets/Scripts/Player/Actor.cs:
```diff
@@ -0,0 +1,117 @@
+        // Make sure there's a movement type specified
+        if(movement == null) {
+            // Throw error if no movement specified for player actor
+            if(PlayerController.Instance.GetActor() == this) {
```
```suggestion
            if(PlayerController.Instance.GetActor() == this) 
            {
```

cptvictor commented on 2D3D_UnityProject/Assets/Scripts/Player/Actor.cs:
```diff
@@ -0,0 +1,117 @@
+            }
+            // Otherwise default to following player
+            // TODO: maybe not a good idea, find a way around this
+            else {
```
```suggestion
            else 
            {
```

cptvictor commented on 2D3D_UnityProject/Assets/Scripts/Player/Actor.cs:
```diff
@@ -0,0 +1,117 @@
+        return moveVector;
+    }
+
+    private void Animation()
```
I usually find it easier to think about functions as actions, change the function name to a verb, something like Animate(), GenerateAnims(), or GenAnims() rather than Animation. Summary section could be beneficial here too.

cptvictor commented on 2D3D_UnityProject/Assets/Scripts/Player/Actor.cs:
```diff
@@ -0,0 +1,117 @@
+        }
+    }
+
+    void Update()
```
```suggestion
    private void Update()
```

cptvictor commented on 2D3D_UnityProject/Assets/Scripts/Player/Actor.cs:
```diff
@@ -0,0 +1,117 @@
+    [SerializeField]
+    protected Movement movement;
+
+    void Start()
```
```suggestion
    private void Start()
```

cptvictor commented on 2D3D_UnityProject/Assets/Scripts/GameManager.cs:
```diff
@@ -3,16 +3,23 @@
+    private PlayerController playerControl;
+
+    public static GameManager instance = null;
     void Awake () {
```
```suggestion
    private void Awake () 
    {
```

cptvictor commented on 2D3D_UnityProject/Assets/Scripts/GameManager.cs:
```diff
@@ -24,36 +31,45 @@ public class GameManager : MonoBehaviour {
 
     void Start () {
```
```suggestion
    private void Start () 
    {
```

cptvictor commented on 2D3D_UnityProject/Assets/Scripts/GameManager.cs:
```diff
@@ -24,36 +31,45 @@ public class GameManager : MonoBehaviour {
+        playerControl = PlayerController.Instance;
     }
 
     void Update () {
```
```suggestion
    private void Update () 
    {
```

cptvictor commented on 2D3D_UnityProject/Assets/Scripts/Player/AnimationController.cs:
```diff
@@ -0,0 +1,63 @@
+    /// </summary>
+    private Vector2 lastFacing;
+
+    void Start()
```
```suggestion
    private void Start()
```

cptvictor commented on 2D3D_UnityProject/Assets/Scripts/Player/AnimationController.cs:
```diff
@@ -0,0 +1,63 @@
+    void Start()
+    {
+        // fuckin sweet way to get components check this shit out
+        if (!TryGetComponent(out anim)) {
```
```suggestion
        if (!TryGetComponent(out anim)) 
        {
```

cptvictor commented on 2D3D_UnityProject/Assets/Scripts/Player/AnimationController.cs:
```diff
@@ -0,0 +1,63 @@
+        if (!TryGetComponent(out anim)) {
+            Debug.LogWarning(name + " | Animator component not found in AnimationController - please attach an Animator to this GameObject");
+        }
+        if (!TryGetComponent(out spriteRenderer)) {
```
```suggestion
        if (!TryGetComponent(out spriteRenderer)) 
        {
```

cptvictor commented on 2D3D_UnityProject/Assets/Scripts/Player/AnimationController.cs:
```diff
@@ -0,0 +1,63 @@
+    public void UpdateAnims(Vector2 movement)
+    {
+        // If player is not moving, use values from last non-zero movement input (so we maintian direction after stopping)
+        if(movement == Vector2.zero) {
```
```suggestion
        if(movement == Vector2.zero) 
        {
```

cptvictor commented on 2D3D_UnityProject/Assets/Scripts/Player/AnimationController.cs:
```diff
@@ -0,0 +1,63 @@
+            movement = lastFacing;
+        }
+        // Otherwise save movement values to maintain direction when stopped
+        else {
```
```suggestion
        else 
        {
```

cptvictor commented on 2D3D_UnityProject/Assets/Scripts/Player/AnimationController.cs:
```diff
@@ -0,0 +1,63 @@
+        anim.SetFloat("Speed", 0);
+    }
+
+    void LateUpdate()
```
```suggestion
    private void LateUpdate()
```

cptvictor commented on 2D3D_UnityProject/Assets/Scripts/Player/Movement/BackViewMovement.cs:
```diff
@@ -0,0 +1,39 @@
+    /// Returns horizontal movement along X-axis
+    /// </summary>
+    /// <param name="player"></param>
+    /// <returns></returns>
```
WHAT DOES IT RETURN!?

cptvictor commented on 2D3D_UnityProject/Assets/Scripts/Player/Movement/BackViewMovement.cs:
```diff
@@ -0,0 +1,39 @@
+
+        // Ignore Up/Down - not used in this movement scheme
+        // Left/Right = change in X
+        if(Input.GetKey(KeyCode.A)) {
```
```suggestion
        if(Input.GetKey(KeyCode.A)) 
        {
```

cptvictor commented on 2D3D_UnityProject/Assets/Scripts/Player/Movement/BackViewMovement.cs:
```diff
@@ -0,0 +1,39 @@
+        if(Input.GetKey(KeyCode.A)) {
+            movement += Vector3.left;
+        }
+        if(Input.GetKey(KeyCode.D)) {
```
```suggestion
        if(Input.GetKey(KeyCode.D)) 
        {
```

cptvictor commented on 2D3D_UnityProject/Assets/Scripts/Player/Movement/BackViewMovement.cs:
```diff
@@ -0,0 +1,39 @@
+        return movement.normalized;
+    }
+
+    public override Vector2 GetAnimation(Actor player)
```
Add summary section

cptvictor reviewed (commented):
Probably a future thing, but shouldn't input be processed by its own class and then sent to the other ones? I.e. Have an input class process when E is pressed and then send the info to the InteractionController rather than have the InteractionController figure out when E is pressed? Could also be easier for implementing rebindable keys if we want that.

Other than that, mostly adding private to Start() functions and putting curly brackets on their own lines.

dmg9626 committed:
refactored GameManager/PerspectiveCameraControl - eliminated redundant ghostCamera gameobject reference

dmg9626 committed:
small cleanup in Actor.cs

dmg9626 committed:
cleanup in Movement classes and Actor/PlayerController

dmg9626 committed:
marked Start() and LateUpdate() as private in AnimationController.cs

dmg9626 committed:
more AnimationController cleanup

dmg9626 committed:
added summary to BackViewMovement.GetAnimation()

dmg9626 committed:
fixed bug: player couldn't exit Zodiac Puzzle

dmg9626 committed:
added NullMovement.cs class - may use this later

dmg9626 committed:
fixed liars puzzle bug; couldn't exit dialouge

(also unified code that enables/disables actor control into Actor.Enable() / Disable())

Commit:
- changed Actor.Update() to FixedUpdate(): move speed no longer tied to framerate

Commit:
reduced actor movement speed from 15 to 10

Commit:
- Actor now supports sprinting (hold SHIFT)

Commit:
removed obsolete/unused/placeholder code

dmg9626 committed:
fixed bug: pressing E when not near interactable broke game

dmg9626 committed:
fixed PlayerInteractionController->interactCanvas nullref when entering teledoor puzzle

dmg9626 committed:
made teledoor colliders larger (increased z-size from 1 to 6, so they protrude more)

dmg9626 committed:
added rigidbody to oliver in teledoor scene; fixed playercontroller->actor nullref in teledoor scene

dmg9626 committed:
changed Oliver actor to use a rigidbody movement system instead of CharacterController (so gravity can work properly) - this fixed the teledoor puzzle

dmg9626 committed:
restored door thickness (collider z-size) from 6 back to 1

n8sthegr8est committed:
fixed the doors and gravity

n8sthegr8est committed:
fixed puzzle and gravity
